### ISSUE TITLE
highly reproducible crash on Windows

### ISSUE THREAD

**Issue by sthalik** [git-cola/git-cola#1191]
To make git-cola crash, press `C-q` immediately upon the program's window appearing. The current working directory must be a repo. pynotify is installed.

**Comment by living180** [comment]
I was able to reproduce a hang in this situation, but not a crash (exception).  Is this the same thing you are seeing?

**Comment by sthalik** [comment]
There's a backtrace:

```
>git cola
Traceback (most recent call last):
  File "C:\Python39\share\git-cola\lib\cola\qtutils.py", line 950, in run
    self.result = self.task()
  File "C:\Python39\share\git-cola\lib\cola\qtutils.py", line 973, in task
    return self.fn(*self.args, **self.kwargs)
  File "C:\Python39\share\git-cola\lib\cola\models\main.py", line 253, in update_status
    self._update_files(update_index=update_index)
  File "C:\Python39\share\git-cola\lib\cola\models\main.py", line 297, in _update_files
    selection.update(self)
  File "C:\Python39\share\git-cola\lib\cola\models\selection.py", line 87, in update
    self.notify_observers(self.message_selection_changed)
  File "C:\Python39\share\git-cola\lib\cola\observable.py", line 30, in notify_observers
    method(*args, **opts)
AttributeError: 'DiffEditor' does not have a signal with the signature updated()
```

**Comment by living180** [comment]
OK, after trying a few more times, I was able to get the same stack trace.  Thanks!

**Comment by living180** [comment]
OK, so after some debugging, it turns out this isn't a Windows-specific problem (though it is easier to reproduce on Windows since Git is slower on that platform) and it doesn't have to do with file change monitoring.  The root of the issue is that when git-cola starts up, it spawns a thread to update the repository state so that the program isn't blocked from starting up if that operation takes a long time (e.g. due to being a large repository):  https://github.com/git-cola/git-cola/blob/74e94ae1fc4299d36adbb1cc456df9a674fa72b3/cola/app.py#L517.

However, if the user quits git-cola before that thread finishes, the thread will end up running `notify_observers()` for observers which then try to run .emit() on signals belonging to Qt objects that have already been destroyed at the C++ level.  On Linux this results in Python exceptions and/or core dumps, depending on the Qt binding used.  Here is an example similar to what @sthalik observed, but using pyside2 on Linux:

    $ QT_API=pyside2 bin/git-cola -r /var/db/repos/gentoo/
    Traceback (most recent call last):
      File "/home/dharding/code/opensource/git-cola/bin/cola/qtutils.py", line 950, in run
        self.result = self.task()
      File "/home/dharding/code/opensource/git-cola/bin/cola/qtutils.py", line 973, in task
        return self.fn(*self.args, **self.kwargs)
      File "/home/dharding/code/opensource/git-cola/bin/cola/models/main.py", line 258, in update_status
        self.update_submodules_list()
      File "/home/dharding/code/opensource/git-cola/bin/cola/models/main.py", line 361, in update_submodules_list
        self.notify_observers(self.message_submodules_changed)
      File "/home/dharding/code/opensource/git-cola/bin/cola/observable.py", line 30, in notify_observers
        method(*args, **opts)
    RuntimeError: Internal C++ object (SubmodulesTreeWidget) already deleted.
    Segmentation fault (core dumped)

The best solution for this problem isn't immediately clear.  The approach I have begun prototyping essentially replaces some usages of the Observer class with using Qt signals directly.  However, this would not be a small change and obviously involves changes to the program design/architecture.  So, before moving forward, I wanted to get feedback from @davvid since I'm not certain of the purpose of Observable vs. Qt signals, and I don't know if this is the best approach, or if there is a better way to solve this.

I pushed a commit demonstrating the approach I am talking about to my fork:  https://github.com/living180/git-cola/tree/poc.  However, fully addressing the crash would require quite a few more changes.  I'm happy to do the work but only if that is the best way forward.

**Comment by sthalik** [comment]
Why not do a blocking wait for the background thread, possibly using `QThread::requestInterruption` to make it do less work? At least that's the path of least resistance here.

**Comment by davvid** [comment]
In general I would prefer using Qt signals/slots over the Observable class these days. That class was written at the time to provide a dead-simple (dependency-free) synchronous notification mechanism and if we can make it better by leveraging Qt then all the better. As you're aware it's a bit embedded in the code so it'd be great to rework that.

There are a few places where we've had to deal with the "C++ object has gone away" problem in the past by catching RuntimeError. The approach in the poc branch looks really nice and I suspect that it may fix the root cause that led to the current try/except spots where we deal with similar RuntimeError scenarios.

The fact that the Observables hold onto widgets in their observer lists was absolutely a friction point in the past. Thanks for diving into that.

**Comment by living180** [comment]
> Why not do a blocking wait for the background thread, possibly using `QThread::requestInterruption` to make it do less work? At least that's the path of least resistance here.

So, git-cola's actual implementation doesn't use `QThread` directly, but instead `QRunnable` and `QThreadPool`.  git-cola already waits for the thread pool during shutdown:

https://github.com/git-cola/git-cola/blob/74e94ae1fc4299d36adbb1cc456df9a674fa72b3/cola/app.py#L401-L403

which is called from line 375 in the following:

https://github.com/git-cola/git-cola/blob/74e94ae1fc4299d36adbb1cc456df9a674fa72b3/cola/app.py#L365-L378

However, with some print debugging I discovered that the exception I saw in my previous comment (https://github.com/git-cola/git-cola/issues/1191#issuecomment-1020429727) was happening before the `.waitForDone()` call, so things are already being torn down before there is a chance for the background task to finish.

So, I'll continue with the approach of replacing Observable and see if that is able to remedy the situation.

**Comment by living180** [comment]
So, I have a (very rough at this point) branch where I have completely eliminated Observable from git-cola:  https://github.com/living180/git-cola/tree/wip.  However, I've run into a bit of a snag.  I'll summarize the issue here, but if there is a better place to discuss, let me know.

I'll start with the symptom:  when running on my branch, if I select the View -> File Browser... menu item to open a file browser window, and then repeatedly expand and collapse the directories in the file browser, eventually the browser window will spontaneously close.  After some arduous debugging, I discovered that this was simply due to the `Browser` instance being garbage collected by Python.

The `Browser` instance is created here https://github.com/git-cola/git-cola/blob/e3b9b4b58dadfe556260cde37a25171ce465551c/cola/widgets/main.py#L540-L542 via the `worktree_browser()` function.  And while `worktree_browser()` returns the `Browser` instance, in this case that reference is immediately discarded once `worktree_browser()` returns.  It turns out that the only remaining reachable reference to the `Browser` instance was due to the `.add_observer()` call in `Browser.__init__`:  https://github.com/git-cola/git-cola/blob/74e94ae1fc4299d36adbb1cc456df9a674fa72b3/cola/widgets/browse.py#L67

However, my changes remove the `.add_observer()` call in favor of a signal connection:  https://github.com/living180/git-cola/commit/5e67d26c17afc4d245069916ffa905ddbf770678#diff-e4f0ceeb629d8dff14a23c45166471a99c1a1345ea1fff92afee16d2f8301810L81-R80.

Thus after my changes there is no longer any reachable reference to the `Browser` reference and it gets evicted from memory once the garbage collector runs.

Architecturally, I am not sure of the best way to solve this.  The best I have come up with is to make a method on `MainView` that calls `worktree_browser()` and then stores the resulting `Browser` on the `MainView`.  `Browser` would then need to override `closeEvent()` to emit a signal.  `MainView` would connect to that signal to remove its reference to the `Browser` once it was closed to avoid leaking memory.  It feels a little clunky, but if no one has any better suggestions, I'll mock it up and add it to my branch.
